File: lc/two-sum.py
#!/usr/bin/env python3
# two-sum.py

from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        d = {}
        for j, num in enumerate(nums):
            i = d.get(target - num, None)
            if i is not None:
                return [i, j]
            d[num] = j
        ValueError("Impossible to have no valid answers!")

if __name__ == "__main__":

    import unittest

    class TestSolution(unittest.TestCase):

        def helper(self, nums: List[int], target: int, want: List[int]):
            got = Solution.twoSum(None, nums, target)
            self.assertEqual(got, want)

        def test_twoSum_example_1(self):
            self.helper([2, 7, 11, 15], 9, [0, 1])

        def test_twoSum_example_2(self):
            self.helper([3, 2, 4], 6, [1, 2])

        def test_twoSum_example_3(self):
            self.helper([3, 3], 6, [0, 1])

    unittest.main()

File: lc/two-sum.rs
use std::collections::HashMap;

struct Solution;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut idxs = HashMap::new();

        for (j, num) in nums.iter().enumerate() {
            let diff = target - num;
            match idxs.get(&diff) {
                None => idxs.insert(num, j as i32),
                Some(i) => return vec![*i, j as i32],
            };
        }

        panic!("Impossible. Guaranteed to have no valid answers.");
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn helper(nums: Vec<i32>, target: i32, want: Vec<i32>) {
        let mut got = Solution::two_sum(nums.clone(), target);
        got.sort();
        eprintln!("two_sum({:?}, {:?}) = {:?}", nums, target, got);
        assert_eq!(got, want);
    }

    #[test]
    fn test_two_sum_1() { helper(vec![2, 7, 11, 15], 9, vec![0, 1]); }

    #[test]
    fn test_two_sum_2() { helper(vec![3, 2, 4], 6, vec![1, 2]); }

    #[test]
    fn test_two_sum_3() { helper(vec![3, 3], 6, vec![0, 1]); }

}
